### PR TITLE
Add and support receiver creator endpoint properties

### PR DESCRIPTION
**Pull request by rmfitzpatrick**
**Description:**
Adding a feature - These changes add a new boolean `accept_endpoint_properties` config field and mechanism to evaluate remotely specified receiver template configuration by observer endpoint environment content (disabled by default). This feature will allow the receiver creator to add or modify evaluated receiver template config, rules, and resource attributes specified by the applicable observer metadata for the containing endpoint alone.

**Link to tracking Issue:**
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/17418

**Testing:**
Added and updated tests suites.

**Documentation:**
Updated readme to describe feature.

### CODE REVIEW

**Comment by dmitryax** [comment]
Couple Qs for clarification:

- What properties do I need to set to scrape metrics only from the annotated (let's say redis) pods and ignore other redis pods that are not annotated?
- Does removing annotations stop metrics scraping?

**Comment by rmfitzpatrick** [comment]
> * What properties do I need to set to scrape metrics only from the annotated (let's say redis) pods and ignore other redis pods that are not annotated?

The rule will only be applied to the endpoint that contains the properties, so it doesn't transfer to any other endpoint evaluation from an unannotated redis pod. This occurs in `observerHandler.OnAdd()` for each observed endpoint.

> * Does removing annotations stop metrics scraping?

Yes, since the `observerHandler.OnChange()` method removes the target receiver with `OnRemove()` and then runs `OnAdd()` with the updated endpoint no longer containing the annotation, its properties aren't able to be evaluated again so only the service config would apply.

**Comment by dmitryax** [comment]
> The rule/config/resource_attributes* will only be applied to the endpoint that contains the properties, so it doesn't transfer to any other endpoint evaluation from an unannotated redis pod. This occurs in observerHandler.OnAdd() for each observed endpoint.

If it's only applied to the endpoint containing the properties, why do we need to transfer the recievercreator's endpoint discovery configuration to the endpoint itself? I'm talking about the `rule` annotation. Shouldn't it be just a boolean annotation or just the presence of any other receiver-creator annotation?

**Comment by rmfitzpatrick** [comment]
> If it's only applied to the endpoint containing the properties, why do we need to transfer the recievercreator's endpoint discovery configuration to the endpoint itself? I'm talking about the `rule` annotation. Shouldn't it be just a boolean annotation or just the presence of any other receiver-creator annotation?

For `port` endpoints, the annotations come from the sub-pod, so a default true would cause a hit for all ports (all container ports), when likely a single is desired. Similarly for `container` endpoints, an endpoint exists for each port ([endpointForPort](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/extension/observer/dockerobserver/extension.go#L150C26-L150C41)). This way specifying the port in the in `type == "port" and port == xyz ` rule is helpful.

**Comment by dmitryax** [comment]
> For port endpoints, the annotations come from the sub-pod, so a default true would cause a hit for all ports (all container ports), when likely a single is desired. Similarly for container endpoints, an endpoint exists for each port ([endpointForPort](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/extension/observer/dockerobserver/extension.go#L150C26-L150C41)). This way specifying the port in the* type == "port" and port == xyz  or type == "container" and port == xyz rules* is helpful.

Shouldn't `port` be provided as part of `config`? Configuring an endpoint discovery rule on the endpoint itself seems a bit confusing.

**Comment by rmfitzpatrick** [comment]
> Shouldn't `port` be provided as part of `config`? Configuring an endpoint discovery rule on the endpoint itself seems a bit confusing.

I think this may be a general receiver creator use question: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/receivercreator#rule-expressions. The `port` variable in an environment can be used in an arbitrary place in config, but is also part of the endpoint identity. Because there is an endpoint generated for each port, all of which share labels/annotations from an env, having the port as an identifier acts as a filter.

**Review by hughesjj — APPROVED** [review]
LGTM

File: receiver/receivercreator/observerhandler.go [hunk 1]
```diff
@@ -182,10 +190,10 @@ func (obs *observerHandler) OnRemove(removed []observer.Endpoint) {
 		}
 
 		for _, rcvr := range obs.receiversByEndpointID.Get(e.ID) {
-			obs.params.TelemetrySettings.Logger.Info("stopping receiver", zap.Reflect("receiver", rcvr), zap.String("endpoint_id", string(e.ID)))
+			obs.logger.Info("stopping receiver", zap.Reflect("receiver", rcvr), zap.String("endpoint_id", string(e.ID)))
```
Comment (atoulme) [inline receiver/receivercreator/observerhandler.go:193]: Probably can peel off this logger change in a separate PR just to make the diff smaller

File: receiver/receivercreator/config.go [hunk 2]
```diff
@@ -135,3 +139,23 @@ func (cfg *Config) Unmarshal(componentParser *confmap.Conf) error {
 
 	return nil
 }
+
+func (rt receiverTemplate) copy() receiverTemplate {
```
Comment (atoulme) [inline receiver/receivercreator/config.go:143]: move it up to around line 70 to keep it close to the struct definition?

**Pull request closed without merging**